### PR TITLE
ISSUE 371 - Mailet to collect contacts

### PR DESCRIPTION
**Pull request by vttranlina**
Resolve https://github.com/linagora/tmail-backend/issues/371

### CODE REVIEW

**Comment by vttranlina** [comment]
Remark: 
- We can't extract contact.firstName, surname from Mail => it will be empty

**Comment by Arsnael** [comment]
> Remark:
> 
>     * We can't extract contact.firstName, surname from Mail => it will be empty

Not convinced on that. I thought you have two syntaxes for recipients in a mail as following:
* `user@host.domain` (what you likely use)
* `Personal Name <user@host.domain>` => name + email... did you try that one?

After how is James reacting to that I'm not sure :)

**Comment by vttranlina** [comment]
> * `Personal Name <user@host.domain>` => name + email... did you try that one?

I see that, but it is a full name, how we can extract it to firstname + surname?
Or we can add one more property (fullName) in ContactFields.

**Comment by Arsnael** [comment]
Hmmm maybe agreeing on an extracting pattern... 

For example split on spaces, first value firstname, other values in surname... or other way around... I don't think it would change much at the end as the search would work more or less the same? 

@chibenwa thoughts?

**Comment by vttranlina** [comment]
Can have a case where each country/region has a different pattern?

**Comment by Arsnael** [comment]
I mean I don't think you can avoid having people putting `Doe John` and others doing `John Doe`. But does it matter? is it gonna change your search at the end as we check all fields anyways? I don't think so honestly

**Comment by vttranlina** [comment]
rebase master

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  private def dispatchEvents(contacts: Seq[ContactFields]): Unit =
+    SFlux.fromIterable(contacts)
+      .flatMap(contact => SMono.fromPublisher(eventBus.dispatch(TmailContactUserAddedEvent(
+        eventId = EventId.random(),
+        username = Username.fromMailAddress(contact.address),
+        contact = contact),
+        NO_REGISTRATION_KEYS)))
+      .collectSeq()
+      .subscribeOn(Schedulers.elastic())
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:55]: subscribeOn then block is a bad practice. It takes 2 thread blocked: the calling one and also ask elastic resources while we could use current thread instead. Please remove it.

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    Seq(Message.RecipientType.TO, Message.RecipientType.CC, Message.RecipientType.BCC)
+      .flatMap(recipientType => Option(mail.getMessage.getHeader(recipientType.toString, ",")))
+      .flatMap(header => InternetAddress.parseHeader(header, false).toSeq)
+      .map(_.toString)
+      .map(MimeUtil.unscrambleHeaderValue)
+      .map(new InternetAddress(_))
+      .map(extractContactField)
+
+  private def extractContactField(internetAddress: InternetAddress): ContactFields  =
+    ContactFields(new MailAddress(internetAddress))
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:45]: Please also use `InternetAddress::getPersonal()` if not null to get one of the name of the dude!

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala [hunk 3]
```diff
@@ -52,11 +52,20 @@ class ContactsCollection @Inject()(eventBus: EventBus) extends GenericMailet {
         contact = contact),
         NO_REGISTRATION_KEYS)))
       .collectSeq()
-      .subscribeOn(Schedulers.elastic())
       .block()
 
   private def appendAttributeToMail(mail: Mail, contacts: Seq[ContactFields]): Unit =
     mail.setAttribute(new Attribute(attributeName,
-      AttributeValue.of(contacts.map(_.address.toString).mkString("[", ",", "]"))))
+      AttributeValue.of(ContactFieldSerializer.serialize(contacts).toString())))
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:59]: No please don't change the format here.

File: tmail-backend/mailets/src/test/scala/com/linagora/tmail/mailets/ContactsCollectionTest.scala [hunk 4]
```diff
@@ -153,8 +154,11 @@ class ContactsCollectionTest {
 
     mailet.service(mail)
 
-    assertThat(getAttributeValue(mail))
-      .hasValue(s"[$RECIPIENT]")
+    val attributeValue = getAttributeValue(mail)
+    assertThat(attributeValue).isPresent
+
+    assertThatJson(attributeValue.get())
+      .isEqualTo(s"""[{"address":"$RECIPIENT","firstname":"","surname":""}]""")
```
Comment (chibenwa) [inline tmail-backend/mailets/src/test/scala/com/linagora/tmail/mailets/ContactsCollectionTest.scala:161]: Please don't change this

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala (hunk 3 above)
Comment (vttranlina) [reply tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:59]: I thought you wanted personal info?

Comment (chibenwa) [reply tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:59]: We want it in our index.

However the attribute value is used to pass the collected contact to the calendar and we need the format of what we send to the calendar to be stable.

If not we will have a bug (non handled message means existing collected contact feature broken) and possibly an outage (I don't trust them managing failed consumers nicely).

Comment (vttranlina) [reply tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala]: I got it

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        eventId = EventId.random(),
+        username = Username.fromMailAddress(contact.address),
+        contact = contact),
+        NO_REGISTRATION_KEYS)))
+      .collectSeq()
+      .block()
+
+  private def appendAttributeToMail(mail: Mail, contacts: Seq[ContactFields]): Unit =
+    mail.setAttribute(new Attribute(attributeName,
+      AttributeValue.of(contacts.map(_.address.toString).mkString("[", ",", "]"))))
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:58]: The format of the attribute should be exactly the one of `ContactExtractor` class in James source code:

```
 * {
 *   "userEmail" : "sender@james.org", 
 *   "emails" : [ "to@james.org", "cc@james.org" ]
 * }
```

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.commons.collections.CollectionUtils
+import org.apache.james.core.{MailAddress, Username}
+import org.apache.james.events.Event.EventId
+import org.apache.james.events.{EventBus, RegistrationKey}
+import org.apache.james.mime4j.util.MimeUtil
+import org.apache.mailet.base.GenericMailet
+import org.apache.mailet.{Attribute, AttributeName, AttributeValue, Mail, MailetException}
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+class ContactsCollection @Inject()(eventBus: EventBus) extends GenericMailet {
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:17]: Missing javadoc

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**